test/server/http_time_tracking_filter_integration_test.cc

```diff
@@ -0,0 +1,147 @@
+    Envoy::IntegrationStreamDecoderPtr response;
+    if (setup_for_upstream_request) {
+      response = sendRequestAndWaitForResponse(request_headers, 0, default_response_headers_, 0);
+    } else {
+      response = codec_client_->makeHeaderOnlyRequest(request_headers);
+      response->waitForEndStream();
+    }
+    return response;
+  }
+
+  const Envoy::Http::TestRequestHeaderMapImpl request_headers_;
+};
+
+class HttpTimeTrackingFilterConfigTest : public testing::Test,
+                                         public Envoy::Event::TestUsingSimulatedTime {};
+
+// WARNING: Don't move this test. Keep it on top.
+// The static stopwatch member of the extension config persists accross tests, and
+// we test the first call to that singleton instance here.
+TEST_F(HttpTimeTrackingFilterConfigTest, GetElapsedNanosSinceLastRequest) {
+  Envoy::Event::SimulatedTimeSystem& time_system = simTime();
+  EXPECT_EQ(Server::HttpTimeTrackingFilterConfig::getElapsedNanosSinceLastRequest(time_system), 0);
+  time_system.setMonotonicTime(1ns);
+  EXPECT_EQ(Server::HttpTimeTrackingFilterConfig::getElapsedNanosSinceLastRequest(time_system), 1);
+}
+
+INSTANTIATE_TEST_SUITE_P(IpVersions, HttpTimeTrackingIntegrationTest,
+                         testing::ValuesIn(Envoy::TestEnvironment::getIpVersionsForTest()));
+
+// Verify expectations with static/file-based time-tracking configuration.
+TEST_P(HttpTimeTrackingIntegrationTest, StaticConfiguration) {
+  setup(fmt::format(kProtoConfigTemplate, kDefaultProtoFragment));
+  Envoy::IntegrationStreamDecoderPtr response = getResponse();
+  int64_t latency;
+  const Envoy::Http::HeaderEntry* latency_header =
+      response->headers().get(Envoy::Http::LowerCaseString(kLatencyResponseHeaderName));
+  ASSERT_NE(latency_header, nullptr);
+  EXPECT_TRUE(absl::SimpleAtoi(latency_header->value().getStringView(), &latency));
+  EXPECT_GT(latency, 0);
+}
+
+// Verify expectations with an empty time-tracking configuration.
+TEST_P(HttpTimeTrackingIntegrationTest, NoStaticConfiguration) {
+  setup(fmt::format(kProtoConfigTemplate, ""));
+  // Don't send any config request header
+  getResponse();
+  // Send a config request header with an empty / default config. Should be a no-op.
+  getResponse("{}");
+  // Send a config request header, this should become effective.
+  Envoy::IntegrationStreamDecoderPtr response =
+      getResponse(fmt::format("{{{}}}", kDefaultProtoFragment));
+  const Envoy::Http::HeaderEntry* latency_header =
+      response->headers().get(Envoy::Http::LowerCaseString(kLatencyResponseHeaderName));
+  ASSERT_NE(latency_header, nullptr);
+  int64_t latency;
+  EXPECT_TRUE(absl::SimpleAtoi(latency_header->value().getStringView(), &latency));
+  // TODO(oschaaf): figure out if we can use simtime here, and verify actual timing matches
+  // what we'd expect using that.
+  EXPECT_GT(latency, 0);
+
+  // Send a malformed config request header. This ought to shortcut and directly reply,
```

Comment:
Can these two parts where we check invalid configuration be their own test cases?